File: src/core/AppController.ts
import { ITTSService } from './interfaces/ITTSService.js';
import { ISettingsManager } from './interfaces/ISettingsManager.js';
import { AudioService } from '../infrastructure/AudioService.js';
import { HotkeyManager } from './HotkeyManager.js';
import { HistoryManager } from './HistoryManager.js';

/**
 * Controller coordinating the core domain services of the TTS application.
 * Acts as the entrypoint coordinator for inputs, command execution, settings adjustments,
 * history lookups, and hotkey actions.
 * Decouples the user interfaces (both CLI and Electron frontend) from underlying logic.
 */
export class AppController {
    /** List of supported synthesis engines. Extensible for future local/cloud providers. */
    private availableModels = ['kokoro'];
    
    /** The active TTS engine. */
    private activeModel = 'kokoro';

    constructor(
        private ttsService: ITTSService,
        private settingsManager: ISettingsManager,
        private audioService: AudioService,
        private hotkeyManager: HotkeyManager,
        private historyManager: HistoryManager
    ) {}

    /**
     * Synthesizes and plays a text input stream.
     * Delegates text synthesis directly to the active TTS provider.
     * @param text The sentence/paragraph text.
     */
    public async processInput(text: string): Promise<void> {
        try {
            await this.ttsService.speak(text);
        } catch (error) {
            console.error('Error processing TTS input:', error);
        }
    }

    /**
     * Parses and routes console/IPC commands starting with a slash (`/`).
     * Supports commands such as /help, /voice, /model, /output, /volume, etc.
     * @param commandLine Raw input command string containing parameters.
     * @returns Promise resolving to true if the program loop should continue, or false to exit.
     */
    public async handleCommand(commandLine: string): Promise<boolean> {
        const parts = commandLine.trim().split(' ');
        const command = parts[0].toLowerCase();
        const args = parts.slice(1);

        switch (command) {
            case '/help':
                this.showHelp();
                break;
            case '/model':
                this.handleModelCommand(args);
                break;
            case '/voice':
                await this.handleVoiceCommand(args);
                break;
            case '/output':
                await this.handleOutputCommand(args);
                break;
            case '/volume':
                await this.handleVolumeCommand(args);
                break;
            case '/monitoring':
                await this.handleMonitoringCommand(args);
                break;
            case '/monitoring_output':
                await this.handleMonitoringOutputCommand(args);
                break;
            case '/volume_monitoring':
                await this.handleVolumeMonitoringCommand(args);
                break;
            case '/history':
                await this.handleHistoryCommand(args);
                break;
            case '/hotkey':
                await this.handleHotkeyCommand(args);
                break;
            case '/exit':
            case '/quit':
                console.log('Exiting.');
                return false; // Signals to close the app
            default:
                console.log('Command not found. Type "/help" for a list of commands.');
                break;
        }
        return true; // Signals to continue running
    }

    /**
     * Outputs help instructions and command usage guidelines to stdout.
     */
    public showHelp(): void {
        console.log(`
Available commands:
  /help                   - Show this help message

  /model --list           - List available TTS models
         <model_name>     - Set the active TTS model
  
  /voice --list           - List available voices for the current model
         <voice_name>     - Set the active voice
  
  /output --list          - List available output devices
          <device_id>     - Set the active output device
  /volume <0.0-1.0>       - Set playback volume (e.g., 0.5 for 50%)
  
  /monitoring <on/off>    - Enable or disable dual output monitoring
  /monitoring_output <id> - Set the monitoring output device
  /volume_monitoring <v>  - Set monitoring output volume (0.0-1.0)
  
  /history                - List the last 20 generated audios
           play <ID>      - Play the specific history file
           delete <ID>    - Delete the specific history file
           clear          - Delete all generated history

  /hotkey list            - List available hotkey combiantions
          play <ID>       - Play a specific hotkeyed phrase
          assign <hotkey> - Assign a hotkey combination to a phrase to generate
          delete <ID>     - Delete a hotkey combination and its phrase
          clear           - Delete all hotkeyed phrases
  
  /exit                   - Quit the application
        `.trim());
    }

    /**
     * Lists registered model identifiers.
     */
    public listModels(): string[] {
        return this.availableModels;
    }

    /**
     * Sets the active voice model.
     * @param modelName Identifier of the model to select.
     */
    public setModel(modelName: string): boolean {
        if (this.availableModels.includes(modelName)) {
            this.activeModel = modelName;
            return true;
        }
        return false;
    }

    /**
     * Gets the currently active model identifier.
     */
    public getActiveModel(): string {
        return this.activeModel;
    }

    /**
     * Handles /model subcommands.
     */
    private handleModelCommand(args: string[]): void {
        if (args.length === 0) {
            console.log('Usage: /model --list OR /model <model_name>');
            return;
        }

        if (args[0] === '--list') {
            console.log(`Available models:\n${this.listModels().join(', ')}`);
        } else {
            const success = this.setModel(args[0]);
            if (success) {
                console.log(`Model set to ${args[0]}.`);
            } else {
                console.log(`Model '${args[0]}' not found.`);
            }
        }
    }

    /**
     * Lists all voice options matching the current synthesis engine.
     */
    public async listVoices(): Promise<string[]> {
        return await this.ttsService.getVoices();
    }

    /**
     * Retrieves the current voice identifier from engine configuration files.
     */
    public async getActiveVoice(): Promise<string> {
        // Hardcoded to kokoro for now since it's the only engine
        const config = await this.settingsManager.getEngineConfig<{voiceId: string}>('kokoro', { voiceId: 'af_heart' });
        return config.voiceId;
    }

    /**
     * Sets a new active voice and persists changes.
     * @param voiceName The identifier of the voice.
     */
    public async setVoice(voiceName: string): Promise<void> {
        await this.ttsService.setVoice(voiceName);
    }

    /**
     * Returns the global application configuration object.
     */
    public async getAppConfig() {
        return await this.settingsManager.getAppConfig();
    }

    /**
     * Patches the current global application configuration and triggers updates.
     * @param config Partial configurations object.
     */
    public async updateAppConfig(config: any) {
        return await this.settingsManager.updateAppConfig(config);
    }

    /**
     * Enumerates output audio devices available on the system.
     */
    public async getDevices() {
        return await this.audioService.getDevices();
    }

    /**
     * Process Voice selection CLI command.
     */
    private async handleVoiceCommand(args: string[]): Promise<void> {
        if (args.length === 0) {
            console.log('Usage: /voice --list OR /voice <voice_name>');
            return;
        }

        try {
            if (args[0] === '--list') {
                const voices = await this.listVoices();
                console.log(`Available voices:\n${voices.join(', ')}`);
            } else {
                await this.setVoice(args[0]);
            }
        } catch (error: any) {
            console.error(error.message);
        }
    }

    /**
     * Process Output device selection CLI command.
     * Translates index number or device ID string to app settings.
     */
    private async handleOutputCommand(args: string[]): Promise<void> {
        if (args.length === 0) {
            console.log('Usage: /output --list OR /output <device_ID>');
            return;
        }

        const devices = await this.audioService.getDevices();

        if (args[0] === '--list') {
            console.log('Available output devices:');
            devices.forEach((device, index) => {
                console.log(`[${index}] ${device.name}`);
            });
        } else {
            const index = parseInt(args[0], 10);
            let selectedId: string | null = null;

            if (!isNaN(index) && index >= 0 && index < devices.length) {
                selectedId = devices[index].id;
                console.log(`Output device set to: ${devices[index].name}`);
            } else {
                // If they passed ID string directly
                const device = devices.find(d => d.id === args[0]);
                if (device) {
                    selectedId = device.id;
                    console.log(`Output device set to: ${device.name}`);
                } else {
                    console.log('Device not found. Use "/output --list" to see valid indices or IDs.');
                    return;
                }
            }

            await this.settingsManager.updateAppConfig({ playbackDevice: selectedId });
        }
    }

    /**
     * Process Volume scaling CLI command.
     */
    private async handleVolumeCommand(args: string[]): Promise<void> {
        if (args.length === 0) {
            console.log('Usage: /volume <0.0-1.0>');
            return;
        }

        const volume = parseFloat(args[0]);
        if (!isNaN(volume) && volume >= 0 && volume <= 2.0) {
            await this.settingsManager.updateAppConfig({ volume });
            console.log(`Volume set to ${Math.round(volume * 100)}%`);
        } else {
            console.log('Invalid volume level. Please use a number between 0.0 and 1.0.');
        }
    }

    /**
     * Process Monitoring toggle CLI command.
     */
    private async handleMonitoringCommand(args: string[]): Promise<void> {
        if (args.length === 0) {
            console.log('Usage: /monitoring <on|off>');
            return;
        }

        const state = args[0].toLowerCase();
        if (state === 'on' || state === 'true' || state === '1') {
            await this.settingsManager.updateAppConfig({ monitoring: true });
            console.log('Monitoring enabled.');
        } else if (state === 'off' || state === 'false' || state === '0') {
            await this.settingsManager.updateAppConfig({ monitoring: false });
            console.log('Monitoring disabled.');
        } else {
            console.log('Usage: /monitoring <on|off>');
        }
    }

    /**
     * Process Monitoring target device selection CLI command.
     */
    private async handleMonitoringOutputCommand(args: string[]): Promise<void> {
        if (args.length === 0) {
            console.log('Usage: /monitoring_output <device_ID> or /monitoring_output --list');
            return;
        }

        const devices = await this.audioService.getDevices();

        if (args[0] === '--list') {
            console.log('Available output devices:');
            devices.forEach((device, index) => {
                console.log(`[${index}] ${device.name} (ID: ${device.id})`);
            });
            return;
        }

        const index = parseInt(args[0], 10);
        let selectedId: string | null = null;

        if (!isNaN(index) && index >= 0 && index < devices.length) {
            selectedId = devices[index].id;
            console.log(`Monitoring output device set to: ${devices[index].name}`);
        } else {
            const device = devices.find(d => d.id === args[0]);
            if (device) {
                selectedId = device.id;
                console.log(`Monitoring output device set to: ${device.name}`);
            } else {
                console.log('Device not found. Use "/monitoring_output --list" to see valid indices or IDs.');
                return;
            }
        }

        await this.settingsManager.updateAppConfig({ monitoringDevice: selectedId });
    }

    /**
     * Process Monitoring volume CLI command.
     */
    private async handleVolumeMonitoringCommand(args: string[]): Promise<void> {
        if (args.length === 0) {
            console.log('Usage: /volume_monitoring <0.0-1.0>');
            return;
        }

        const volume = parseFloat(args[0]);
        if (!isNaN(volume) && volume >= 0 && volume <= 2.0) {
            await this.settingsManager.updateAppConfig({ monitoringVolume: volume });
            console.log(`Monitoring volume set to ${Math.round(volume * 100)}%`);
        } else {
            console.log('Invalid volume level. Please use a number between 0.0 and 1.0.');
        }
    }

    /**
     * Process History command and subcommands (list, clear, play, delete).
     */
    private async handleHistoryCommand(args: string[]): Promise<void> {
        if (args.length > 0) {
            const subCommand = args[0].toLowerCase();
            
            if (subCommand === 'clear') {
                await this.historyManager.clearHistory();
                return;
            } else if (subCommand === 'play' && args.length > 1) {
                const id = parseInt(args[1], 10);
                if (isNaN(id) || id < 0 || id > 19) {
                    console.log('Invalid history ID.');
                    return;
                }
                await this.historyManager.playHistory(id);
                return;
            } else if (subCommand === 'delete' && args.length > 1) {
                const id = parseInt(args[1], 10);
                if (isNaN(id) || id < 0 || id > 19) {
                    console.log('Invalid history ID.');
                    return;
                }
                await this.historyManager.deleteHistory(id);
                return;
            } else {
                console.log('Usage: /history OR /history clear OR /history play <ID> OR /history delete <ID>');
                return;
            }
        }

        // Default: List history
        await this.historyManager.printHistory();
    }

    /**
     * Process Hotkey commands (list, assign, delete, play, clear).
     */
    private async handleHotkeyCommand(args: string[]): Promise<void> {
        if (args.length === 0) {
            console.log('Usage: /hotkey list OR /hotkey play <ID> OR /hotkey assign <hotkey> <text...>');
            return;
        }

        const subCommand = args[0].toLowerCase();

        if (subCommand === 'list') {
            this.hotkeyManager.printHotkeys();
        } else if (subCommand === 'play' && args.length > 1) {
            const id = parseInt(args[1], 10);
            if (isNaN(id)) {
                console.log('Invalid hotkey ID.');
            } else {
                await this.hotkeyManager.playHotkey(id);
            }
        } else if (subCommand === 'delete' && args.length > 1) {
            const id = parseInt(args[1], 10);
            if (isNaN(id)) {
                console.log('Invalid hotkey ID.');
            } else {
                await this.hotkeyManager.deleteHotkey(id);
            }
        } else if (subCommand === 'clear') {
            await this.hotkeyManager.clearHotkeys();
        } else if (subCommand === 'assign' && args.length > 2) {
            const hotkey = args[1];
            const text = args.slice(2).join(' ');
            await this.hotkeyManager.assignHotkey(hotkey, text);
        } else {
            console.log('Usage: /hotkey list OR /hotkey clear OR /hotkey play <ID> OR /hotkey delete <ID> OR /hotkey assign <hotkey> <text...>');
        }
    }
}

File: src/core/app_controller.py
"""
Application controller coordinating TTS, settings, audio, history, and hotkeys.
Port of the TypeScript AppController.ts.
"""
import logging
from typing import Any, Optional

logger = logging.getLogger(__name__)


from src.core.dictionary_manager import DictionaryManager

class AppController:
    """
    Facade coordinating the core domain services of the TTS application.
    Decouples UI layers (both CLI and Qt frontend) from underlying business logic.
    """

    def __init__(
        self,
        tts_service: Any,
        settings_manager: Any,
        audio_service: Any,
        hotkey_manager: Any,
        history_manager: Any,
        model_manager: Any = None
    ) -> None:
        """Initialise the application controller.

        Args:
            tts_service: The core Text-to-Speech provider.
            settings_manager: System for reading/writing configuration.
            audio_service: Audio playback subsystem.
            hotkey_manager: Service for handling global shortcut actions.
            history_manager: Logging system for synthesized text history.
            model_manager: Handles downloading and managing TTS models.
        """
        self._tts_service = tts_service
        self._settings_manager = settings_manager
        self._audio_service = audio_service
        self._hotkey_manager = hotkey_manager
        self._history_manager = history_manager
        self._model_manager = model_manager
        self._dictionary_manager = DictionaryManager(settings_manager)
        self._available_models = self.list_models()
        saved_active = self._settings_manager.get_app_config().get("activeModel", "")
        if saved_active in self._available_models:
            self._active_model = saved_active
        else:
            self._active_model = self._available_models[0] if self._available_models else ""

    # =========================================================================
    # TTS
    # =========================================================================

    def process_input(self, text: str) -> None:
        """Synthesize and play back the given text."""
        try:
            # Apply dictionary replacements
            processed_text = self._dictionary_manager.replace_text(text)
            self._tts_service.speak(processed_text)
        except Exception as e:
            logger.error(f"Error processing TTS input: {e}")

    # =========================================================================
    # Models
    # =========================================================================

    def list_models(self) -> list[str]:
        """Return available TTS engine identifiers with precision suffix."""
        if hasattr(self._tts_service, "is_available") and self._tts_service.is_available():
            if self._model_manager and hasattr(self._model_manager, "get_installed_precisions"):
                precisions = self._model_manager.get_installed_precisions()
                if precisions:
                    return [f"kokoro_{p}" for p in precisions]
            config = self._settings_manager.get_app_config()
            precision = config.get("kokoroPrecision", "")
            name = f"kokoro_{precision}" if precision else "kokoro"
            return [name]
        return []

    def get_active_model(self) -> str:
        """Return the currently selected model name if available, else empty string."""
        active = self._active_model
        if active in self.list_models():
            return active
        return ""

    def set_model(self, model_name: str) -> bool:
        """Set the active model if it exists."""
        if model_name in self.list_models():
            self._active_model = model_name
            self._settings_manager.update_app_config({"activeModel": model_name})
            if hasattr(self._tts_service, "reload_model"):
                self._tts_service.reload_model()
            return True
        return False

    def reload_engine(self) -> None:
        """Re-check engine availability and preload if a model is now present."""
        if hasattr(self._tts_service, "is_available") and self._tts_service.is_available():
            try:
                if hasattr(self._tts_service, "reload_model"):
                    self._tts_service.reload_model()
                else:
                    self._tts_service.preload_model()
                models = self.list_models()
                saved_active = self._settings_manager.get_app_config().get("activeModel", "")
                if saved_active in models:
                    self._active_model = saved_active
                elif models:
                    self._active_model = models[0]
                logger.info(f"Engine reloaded. Active model: {self._active_model}")
            except Exception as e:
                logger.error(f"Failed to reload engine: {e}")

    # =========================================================================
    # Voices
    # =========================================================================

    def list_voices(self) -> list[str]:
        """Return voice IDs from the active TTS provider."""
        if hasattr(self._tts_service, "is_available") and not self._tts_service.is_available():
            return []
        try:
            return self._tts_service.get_voices()
        except Exception as e:
            logger.error(f"Failed to list voices: {e}")
            return []

    def get_active_voice(self) -> str:
        """Return the currently active voice ID from engine config."""
        try:
            config = self._settings_manager.get_engine_config("kokoro", {"voiceId": "af_heart"})
            return config.get("voiceId", "af_heart")
        except Exception:
            return "af_heart"

    def set_voice(self, voice_id: str) -> None:
        """Set the active voice."""
        try:
            self._tts_service.set_voice(voice_id)
        except Exception as e:
            logger.error(f"Failed to set voice: {e}")

    def get_speed(self) -> float:
        """Return the synthesis speed from Kokoro configuration."""
        try:
            config = self._settings_manager.get_engine_config("kokoro", {"speed": 1.0})
            return float(config.get("speed", 1.0))
        except Exception:
            return 1.0

    def set_speed(self, speed: float) -> None:
        """Set the synthesis speed in Kokoro configuration."""
        try:
            self._settings_manager.update_engine_config("kokoro", {"speed": speed})
            if hasattr(self._tts_service, "kokoro_config"):
                self._tts_service.kokoro_config["speed"] = speed
        except Exception as e:
            logger.error(f"Failed to set speed: {e}")

    # =========================================================================
    # Settings
    # =========================================================================

    def get_app_config(self) -> dict:
        """Return the current application configuration."""
        return self._settings_manager.get_app_config()

    def update_app_config(self, config: dict) -> None:
        """Update partial app configuration."""
        self._settings_manager.update_app_config(config)

        if "openOnStartup" in config:
            import sys
            if sys.platform == "win32":
                self._handle_windows_startup(config["openOnStartup"])

    def _handle_windows_startup(self, enable: bool) -> None:
        try:
            import winreg
            import sys
            from pathlib import Path
            
            key = winreg.OpenKey(
                winreg.HKEY_CURRENT_USER, 
                r"Software\Microsoft\Windows\CurrentVersion\Run", 
                0, 
                winreg.KEY_SET_VALUE | winreg.KEY_QUERY_VALUE
            )
            
            app_name = "MoonTTS"
            
            if enable:
                main_script = Path(sys.argv[0]).resolve()
                if main_script.suffix == ".py":
                    cmd = f'"{sys.executable}" "{main_script}"'
                else:
                    cmd = f'"{sys.executable}"'
                
                winreg.SetValueEx(key, app_name, 0, winreg.REG_SZ, cmd)
            else:
                try:
                    winreg.DeleteValue(key, app_name)
                except FileNotFoundError:
                    pass
            winreg.CloseKey(key)
        except Exception as e:
            logger.error(f"Failed to set Windows startup registry: {e}")

    # =========================================================================
    # Audio Devices
    # =========================================================================

    def get_devices(self) -> list[dict]:
        """Return available audio output devices."""
        return self._audio_service.get_devices()

    # =========================================================================
    # History
    # =========================================================================

    def get_history(self) -> list[str]:
        """Return synthesis history texts."""
        return self._history_manager.get_history()

    def play_history(self, history_id: int) -> None:
        """Play a historical recording."""
        self._history_manager.play_history(history_id)

    def delete_history(self, history_id: int) -> None:
        """Delete a history entry."""
        self._history_manager.delete_history(history_id)

    def clear_history(self) -> None:
        """Clear all history."""
        self._history_manager.clear_history()

    # =========================================================================
    # Hotkeys
    # =========================================================================

    def set_hotkeys_changed_callback(self, callback) -> None:
        """Set a callback to be invoked when hotkeys change."""
        self._hotkey_manager.set_hotkeys_changed_callback(callback)

    def list_hotkeys(self) -> list[dict]:
        """Return assigned hotkey entries."""
        return self._hotkey_manager.list_hotkeys()

    def assign_hotkey(self, hotkey: str, text: str) -> None:
        """Assign a hotkey to a phrase."""
        self._hotkey_manager.assign_hotkey(hotkey, text)

    def play_hotkey(self, hotkey_id: str) -> None:
        """Play a saved hotkey phrase."""
        self._hotkey_manager.play_hotkey(hotkey_id)

    def delete_hotkey(self, hotkey_id: str) -> None:
        """Delete a saved hotkey phrase."""
        self._hotkey_manager.delete_hotkey(hotkey_id)

    def clear_hotkeys(self) -> None:
        """Clear all hotkeys."""
        self._hotkey_manager.clear_hotkeys()

    # =========================================================================
    # Dictionary
    # =========================================================================

    def get_dictionary(self) -> list[dict]:
        return self._dictionary_manager.get_dictionary()

    def add_dictionary_entry(self, original: str, spelling: str, case_sensitive: bool) -> bool:
        return self._dictionary_manager.add_entry(original, spelling, case_sensitive)

    def update_dictionary_entry(self, index: int, original: str, spelling: str, case_sensitive: bool) -> bool:
        return self._dictionary_manager.update_entry(index, original, spelling, case_sensitive)

    def delete_dictionary_entry(self, index: int) -> bool:
        return self._dictionary_manager.delete_entry(index)

    def preview_dictionary_spelling(self, spelling: str) -> None:
        if hasattr(self._tts_service, "preview_spelling"):
            self._tts_service.preview_spelling(spelling)
        else:
            logger.warning("preview_spelling not supported by active TTS service.")



    # =========================================================================
    # Model Management
    # =========================================================================

    def get_model_manager(self) -> Any:
        return self._model_manager

    def check_tts_engine_status(self) -> bool:
        """Check if TTS engine is available."""
        if hasattr(self._tts_service, "is_available"):
            return self._tts_service.is_available()
        return True
